### PR TITLE
Add royalities

### PR DESCRIPTION
**Pull request by jfischoff**

### CODE REVIEW

**Comment by jfischoff** [comment]
I haven't tested this but it compiles.

File: src/Market/Offchain.hs [hunk 1]
```diff
@@ -67,7 +67,14 @@ startSale sp = do
     pkh <- pubKeyHash <$> Contract.ownPubKey
     utxos <- utxosAt (pubKeyHashAddress pkh)
     let val     = Value.singleton (sCs sp) (sTn sp) 1
-        dat     = NFTSale { nSeller = pkh, nToken = sTn sp, nCurrency = sCs sp, nPrice = sPrice sp }
+        dat     = NFTSale
+          { nSeller = pkh
+          , nToken = sTn sp
+          , nCurrency = sCs sp
+          , nPrice = sPrice sp
+          , nRoyalityAddress = sRoyalityAddress sp
```
Comment (BlakeBrown) [inline src/Market/Offchain.hs:75]: Shouldn't this also be a pkh?

`nRoyalityPkh` just for consistency?

File: src/Market/Types.hs [hunk 2]
```diff
@@ -28,18 +28,22 @@ import           Plutus.Contract           ( Endpoint, type (.\/) )
...
 data NFTSale = NFTSale
-    { nSeller    :: !PubKeyHash
-    , nPrice     :: !Integer
-    , nCurrency  :: !CurrencySymbol
-    , nToken     :: !TokenName
+    { nSeller          :: !PubKeyHash
+    , nPrice           :: !Integer
+    , nCurrency        :: !CurrencySymbol
+    , nToken           :: !TokenName
+    , nRoyalityAddress :: !PubKeyHash
```
Comment (BlakeBrown) [inline src/Market/Types.hs]: `nRoyalty` just to be clear it's a pkh, not exactly an address

File: src/Market/Types.hs [hunk 3]
```diff
@@ -28,18 +28,22 @@ import           Plutus.Contract           ( Endpoint, type (.\/) )
...
+    , nRoyalityPercent :: !Integer
```
Comment (BlakeBrown) [inline src/Market/Types.hs]: `nRoyaltyPercent` you have a typo :p

Comment (BlakeBrown) [reply src/Market/Types.hs]: We should also support percentages up to two decimal places, whether you want to multiply this by 100 in the validator or change the type both are okay 😄

File: src/Market/Onchain.hs [hunk 4]
```diff
@@ -77,18 +78,28 @@ mkBuyValidator pkh nfts r ctx =
...
+
+    marketplaceFees :: Integer
+    marketplaceFees = 20
+
+    sellerPercentage :: Integer
+    sellerPercentage = (1000 - marketplaceFees) - nRoyaltyPercent nfts
+
     checkSellerOut :: Bool
-    checkSellerOut = fromInteger (Ada.getLovelace (Ada.fromValue (valuePaidTo info seller))) >= 98 % 100 * fromInteger price
+    checkSellerOut = fromInteger (Ada.getLovelace (Ada.fromValue (valuePaidTo info seller))) >= sellerPercentage % 1000 * fromInteger price
```
Comment (BlakeBrown) [inline src/Market/Onchain.hs]: We need to `checkSellerOut` as the last step by subtracting the fees & royalties from the previous steps.

Ex. Selling an asset for 10 ADA
max(1 ADA, marketplaceFee) will require 1 ADA from the purchase price
max(1 ADA, royaltyFee) will require 1 ADA from the purchase price, even if royalty is 0% (bug 1)
sellerReceives 8 ADA, but the script is expecting them to receive 10 * 0.98 = 9.80 ADA (bug 2)

**Comment by jfischoff** [comment]
Okay, that makes sense.

**Comment by BlakeBrown** [comment]
Thank you Jonathan! 🎈

**Pull request merged**